src/storage/fileio/unix/mod.rs

```diff
@@ -0,0 +1,195 @@
+use nix::libc;
+use std::fs::OpenOptions;
+use std::os::unix::prelude::FileExt;
+use std::{fs::File, io};
+
+use super::FileIO;
+
+pub struct UnixFileIO {}
+
+impl UnixFileIO {
+    #[cfg(target_os = "macos")]
+    fn open_or_create_file(path: String, write: bool, create: bool) -> io::Result<File> {
+        use std::os::unix::prelude::AsRawFd;
+
+        let file = OpenOptions::new()
+            .read(true)
+            .write(write)
+            .create(create)
+            .open(&path);
+
+        if file.is_err() {
+            return file;
+        }
+
+        let file_fd = file.as_ref().unwrap().as_raw_fd();
+        // On OSX, there is no O_DIRECT flag to avoid cache, so we use
+        // alternative F_NOCACHE that is fnctl called on the fd
+        unsafe { libc::fcntl(file_fd, libc::F_NOCACHE, 1) };
+
+        file
+    }
+
+    #[cfg(not(target_os = "macos"))]
+    fn open_or_create_file(path: String, write: bool, create: bool) -> io::Result<File> {
```

Comment:
Same as above. You can make `UnixFileIO` a subtrait of `FileIO` and default implement these functions for any `T: UnixFileIO` and leave `open_or_create_file` as the only function required for `MacFileIO`/`LinuxFileIO`.

Reply:
Both of those could still be placed into this file.